<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет Глеб! Меня зовут Марат, и я буду твоим ревьюером. Спешу сообщить что все ключевые этапы в работе выполнены,  с задачей тебе удалось справиться. По поводу обращения - в IT сфере принято общаться на «ты» :) Но, если привычней на «вы», дай знать. Как ревьюера моя задача помочь тебе в развитии, дав хорошие советы. Я внимательно посмотрю твой код, ознакомлюсь с твоими выводами и оставлю комментарии. Где то могу предложить небольшие исправление в коде, но ненавязчиво. Где потребуются уточнения, я оставлю много наводящих вопросов. Они помогут тебя с поиском верного решения.

Все мои комментарии размечены по цветам, для лучшего восприятия. 
    
<div class="alert alert-success">Зеленым цветом и словом «Успех» отмечены особо удачные и элегантные решения, которыми ты можешь гордиться. </div>
        
<div class="alert alert-warning">Желтым и значком словом «Совет», помечены решения у которых есть альтернативные решения, более оптимальные. Ты можешь найти их сразу и доработать проект, или отложить это на потом, для будущих проектах. Проект будет принят и без их доработки. </div>
        
<div class="alert alert-danger"> Красным цветом и значком словом «Ошибка» помечу твои решения, на которые стоит обратить внимание прежде всего. После их доработки проект будет принят. </div>
        
Залог успеха - работа сообща, взаимное уважение и работа в диалоге. Поэтому, помечай свои ответные комментарии на мои реплики заметным цветом или курсивом, так мне будет легче их отслеживать. Пожалуйста, не изменяй и не удаляй мои комментарии. Все это поможет выполнить повторную проверку быстрей.

</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Вступление в работу очень важно, так человек, который смотрит твой проект (и на работе в том числе) будет сразу введен в курс дела. 
     
    
 

<div class="alert alert-warning">


Совет: 
   
Вопросик, при желании можешь ответить )
    
    
- а почему по твоему была выбрана метрика f1? 
    
    
- а что если бы нам было нужно найти как можно больше токсичных комментариев, в этом случаи на какую метрику мы бы ориентировались?
    
    
- каким образом мы можем изменить функцию ошибки в модели, чтобы она максимизировала интересующую нас метрику (accuracy, f1, precision, roc-auc итп)?    



</div>

## Подготовка

In [1]:
import numpy as np
import pandas as pd
import re
import nltk
nltk.download('wordnet')
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer 
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression, SGDClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score
import warnings
warnings.filterwarnings('ignore')

[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

- Собираем все импорты в верхней части, чтобы легче было ориентироваться и добавлять новые по необходимости. 



<div class="alert alert-warning">

Совет:     



- кстати есть рекомендации PEP-8 при написании кода, в том числе и для импортов. Если интересно можешь почитать [тут](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html), это на будущее


<div>



In [2]:
try:
    data_toxic_comments=pd.read_csv( 'C:/Users/elve/Desktop/Обучение/data/toxic_comments.csv', sep=',')
except:
    data_toxic_comments=pd.read_csv('/datasets/toxic_comments.csv')


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Молодец что используешь конструкцию try - except (if - else), этим ты проявляешь уважение к тем кто будет работать с твоим кодом.

</div>

In [3]:
data_toxic_comments['toxic'].value_counts()/data_toxic_comments.shape[0]

0    0.898388
1    0.101612
Name: toxic, dtype: float64

In [4]:
def check_data (data_i):
    temp = data_i.copy() 
    # проверяем дубликаты
    print('Кол-во дубликатов:',temp.duplicated().sum())
    list_c=data_i.columns
    for col_l in list_c:
        print('-------------------------')
        #print(col_l, temp[col_l].sort_values().unique())
        print(col_l,'|кол-во строк',len(temp),'| кол-во NaN',temp[col_l].isna().sum(),
        '| процент NaN', round(temp[col_l].isna().sum()/len(temp)*100, 2),'%')

In [5]:
check_data (data_toxic_comments)

Кол-во дубликатов: 0
-------------------------
Unnamed: 0 |кол-во строк 159292 | кол-во NaN 0 | процент NaN 0.0 %
-------------------------
text |кол-во строк 159292 | кол-во NaN 0 | процент NaN 0.0 %
-------------------------
toxic |кол-во строк 159292 | кол-во NaN 0 | процент NaN 0.0 %


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

- Данные изучены. Небольшой EDA не помешает, так как это аналитический проект. 

    

    
-  проверку на сбалансированность 




<div class="alert alert-warning">

Совет: 

    
- describe(), info(), .sample() не помешали бы
    
    
    
    
- нет никакого промужуточного вывода в конце раздела, и никак он не выделен    



- на несбалансированность обратил внимание, это хорошо, но у тебя во всем проекте нет ни одного графика, а ведь красивый, хорошо оформленный график может быть украшением проекта. Почему бы тут его не использовать?   
    
    







</div>




In [6]:
def get_wordnet_pos(word):
    """Map POS tag to first character lemmatize() accepts"""
    tag = nltk.pos_tag([word])[0][1][0].upper()
    print(tag)
    tag_dict = {"J": wordnet.ADJ,
                "N": wordnet.NOUN,
                "V": wordnet.VERB,
                "R": wordnet.ADV}
    return tag_dict.get(tag, wordnet.NOUN)


In [7]:
%%time


def lemmatize_text(text):
   
    lemmatizer = WordNetLemmatizer() 
    text = text.lower()
    text_re = re.sub(r'[^a-zA-Z]', ' ', text)
    text_split=text_re.split()
    lemm_text = [lemmatizer.lemmatize(word,get_wordnet_pos(word)) for word in text_split]
    return " ".join(lemm_text)

data_toxic_comments['lemm_text'] = data_toxic_comments['text'].apply(lemmatize_text)




CPU times: user 33.4 s, sys: 124 ms, total: 33.5 s
Wall time: 33.5 s


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:

- WordNetLemmatizer  рабочий вариант, но у него есть особенности, для корректной работы ему нужно передавать не просто слово, но и POS-тег (Part of Speech, части речи). Набираемся ума-разума [тут](https://webdevblog.ru/podhody-lemmatizacii-s-primerami-v-python/) ) 



<div class="alert alert-warning">


Совет: 


    
- после очистки и лемитизации можно провести частотный анализ текста/[облако слов](https://habr.com/ru/post/517410/) - чтобы получить общее представление о тематике и о наиболее часто встерчаемых словах Кроме того графики, рисунки делают проект визуально интересней
    
    




</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b>
не понимаю как применить POS.что я делаю не так?
</div>

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Совет 🤔:

У меня все сработало

    
    
</div>


In [8]:
data_toxic_comments.head()

Unnamed: 0                                               text  toxic  \
0           0  Explanation\nWhy the edits made under my usern...      0   
1           1  D'aww! He matches this background colour I'm s...      0   
2           2  Hey man, I'm really not trying to edit war. It...      0   
3           3  "\nMore\nI can't make any real suggestions on ...      0   
4           4  You, sir, are my hero. Any chance you remember...      0   

                                           lemm_text  
0  explanation why the edits made under my userna...  
1  d aww he match this background colour i m seem...  
2  hey man i m really not trying to edit war it s...  
3  more i can t make any real suggestion on impro...  
4  you sir are my hero any chance you remember wh...



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Да, всегда лучше проверить что получилось  в итоге, так всегда будет возможность поправить ошбку


</div>


In [9]:
data_toxic_comments = data_toxic_comments.drop(['text'], axis=1)


In [10]:
target = data_toxic_comments['toxic']
features = data_toxic_comments.drop(['toxic'], axis=1)

features_train, features_valid, target_train, target_valid = train_test_split(features, 
                                                                              target, 
                                                                              test_size=0.4, 
                                                                              random_state=1234567)
features_valid, features_test, target_valid, target_test = train_test_split(features_valid, 
                                                                            target_valid, 
                                                                            test_size=0.5,
                                                                            random_state=1234567)

nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))

count_tf_idf = TfidfVectorizer(stop_words=stopwords)

features_train = count_tf_idf.fit_transform(features_train['lemm_text'])
features_valid = count_tf_idf.transform(features_valid['lemm_text'])
features_test = count_tf_idf.transform(features_test['lemm_text'])


[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

train_test_split
    
    
- random_state на месте



<div class="alert alert-warning">



Совет: 

- Ниже ты используешь GridSearchCV, вопрос - зачем тебе валидационная выборка?   GS разобьет на выборки сам, а результат на валидационной можно посмотреть через .best_score_. Хорошо бы обьяснить зачем так делаешь



- Обрати внимание на аргумент stratify, он позволит сохранить изначальное распределение таргетов во всех новых датасетах.  Существующий дисбаланс никуда не денется, но в каждом датасете он будет одинаковым.  Это важно потому что великий рандом может нам разбить датасеты так, что модель получит искаженное представление пропорциях единичек и нулей в таргете    




<div>



<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:


values.astype('U')
    
    
Да, в тренажере был текст на кирилице, там перевод в unicode оправдан. В нашем случае (латиница) это лишь  увеличит количество потребляемой памяти и это в лучшем случаи, в худшем он обрушает ядро.




</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b>
убрал values.astype('U')
</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Успех 👍:

👍
</div>


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Не забыли о стопсловах, они ни к чему и код побежит быстрей

    
<div class="alert alert-warning">


Совет:     

Вопросик:

А стопслова важней убирать  когда мы используем TF-IDF, или когда используе обычный CountVectorizer? 



</div>


### Обучение

In [11]:
%%time

model = LogisticRegression(random_state=1234567)
hyperparams = [{'solver':['newton-cg', 'lbfgs', 'liblinear'],'C':[0.1, 1, 10],'random_state':[1234567]}]
model_g = GridSearchCV(model, hyperparams, scoring='f1')
model_g.fit(features_train, target_train)
print("Лучшие параметры")
print(model_g.best_params_)


Лучшие параметры
{'C': 10, 'random_state': 1234567, 'solver': 'newton-cg'}
CPU times: user 4min 19s, sys: 8min 16s, total: 12min 35s
Wall time: 12min 36s


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:

    
Не забываем random_state при инициализации модели, и тут и далее обрати внимание    
    
</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV3</b></font>

Успех 👍:

👍
</div>




<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>
    
Совет: 




Молодец что используешь GrisSearch, но еще лучше использовать связку GridSearchCV + pipeline. 


О pipeline:

[Pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html), это тема которая сразу затрагивает кроссвалидацию, тюнинг "векторайз", подбор гиперпараметров модели и о том что код стоит делать компактным.
    
    
- в TfidfVectorizer(stop_words=stopwords) у тебя по умолчанию ngram_range=(1, 1), тут можно подбирать разное число n- грамм (и другие параметры), максимизируя метрику, но как объединить перебор по ngram_range с обучением моделей, чтобы не делать это по отдельности или с использованием цикла?! pipeline! Готовый [пример для работы с текстами](https://scikit-learn.org/stable/auto_examples/model_selection/grid_search_text_feature_extraction.html)
    
    
- как избежать ошибки подглядывания в будущее, когда мы предварительно работаем с данными (шкалирование, нормализация, TfidfVectorizer итп итд)? pipeline! особенно это важно, когда мы используем кроссвалидацию. Для TfidfVectorizer делаем .fit (обучаемся) на train, а transform на test, но точно также нужно сделать для валидационной выборки. Но GS делает валидационные внутри себя, спрашивается как добраться до нее и избежать подглядывания в будущее? Казалось бы никак, но нет! Pipeline! ) 
    
    
- pipeline позволяет делать наш код компактней и читабельней, это большой плюс, когда код будет раздуваться     
    
    

         
Если раньше не использовалА pipeline то могу посоветовать видео в котором [индус](https://www.youtube.com/watch?v=mOYJCR0IDk8&ab_channel=HimanshuChandra) на английском с сильным акцентом, но на пальцах обьясняет  самое непонятное (по моему опыту): сопряженность методов fit и transform. Там же есть и код и ссылка на текст. Мне помогло )



В общем если сделать GS+pipeline будет вообще хорошо )  
    
<div>   

In [12]:
%%time

model = LogisticRegression(random_state=1234567)
model.set_params(**model_g.best_params_)
model.fit(features_train, target_train)
target_predict = model.predict(features_valid)
f1_LR = f1_score(target_valid, target_predict)
print('F1 на валидации', f1_LR)

F1 на валидации 0.7722087798133425
CPU times: user 8.61 s, sys: 14 s, total: 22.6 s
Wall time: 22.7 s


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Совет:

    
Этот блок лишний.
    
- ты используешь GS как инструмент получить лучшие значения гиперпараметров, после чего вставляешь их в модель и опять обучаешь. Но GS уже это сделал, готовая моделька тут: .best_estimator_

    
- И валидационный датасет не нужен, потому что GS когда искал лучшие значения гиперпараметров, лучшее значение метрики на валидационном датасете (на самом деле нескольких, ведь кроссвалидация) сохранил в .best_score_    
    
</div>

In [13]:
%%time

model = DecisionTreeClassifier(random_state=1234567)
hyperparams = [{'max_depth':[x for x in range(10,50,5)], 'random_state':[1234567]}]
model_grid = GridSearchCV(model, hyperparams, scoring='f1')
model_grid.fit(features_train, target_train)
print("Лучший результат:")
print(model_grid.best_params_)


Лучший результат:
{'max_depth': 45, 'random_state': 1234567}
CPU times: user 8min 9s, sys: 15.2 s, total: 8min 24s
Wall time: 8min 25s


In [14]:
%%time

model = DecisionTreeClassifier(random_state=1234567)
model.set_params(**model_grid.best_params_)
model.fit(features_train, target_train)
target_predict = model.predict(features_valid)
f1_DTC = f1_score(target_valid, target_predict)
print('F1 на валидации', f1_DTC)

F1 на валидации 0.6960242467463006
CPU times: user 16.8 s, sys: 327 ms, total: 17.2 s
Wall time: 17.2 s


In [15]:
rezult = pd.DataFrame({ 'Параметры': ['F1'],
                       'LogisticRegression':[f1_LR], 'DecisionTreeClassifier':[f1_DTC]})
rezult.head()

Параметры  LogisticRegression  DecisionTreeClassifier
0        F1            0.772209                0.696024

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Совет:

Наглядно и здорово что оформил в табличку   
    
    
</div>





### Тестирование

In [16]:
%%time
model = LogisticRegression(random_state=1234567)
model.set_params(**model_g.best_params_)
model.fit(features_train, target_train)
target_predict = model.predict(features_test)
f1_LR = f1_score(target_test, target_predict)
print('F1 на валидации', f1_LR)

F1 на валидации 0.7656652360515022
CPU times: user 8.35 s, sys: 14.5 s, total: 22.8 s
Wall time: 22.8 s


<b>Вывод:</b>  

В данном проекте был рассмотрен способ нормализации текста, для дальнейшего использования в ML моделях.  

При подготовке проекта был использован метод лемматизации и векторизации корпуса текста.  

После было проведено обучение на двух классических моделях: LogisticRegression и DecisionTreeClassifier.  
Модель LR показала наилучший результат по ключевой метрике и была взята в работу на тестовой выборке.
На тестовой выборке по метрике F1 модель LogisticRegression() дала необходимый результат 0.77.


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Успех 👍:

Кратенько добавил )
</div>


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех: 

- Все верно, логика моделирования не нарушена, тут тестируем только лучшую модель отобранную на валидации, или парочку лучших, если на валидации результаты близки


- Если студент получил на тесте f1 выше 0,75, это считается приемлемым результатом.


<div class="alert alert-warning">


Совет: 

Что может помочь добиться лучшего результата (от простого)? 

- использовать stratify. 
    



- учесть дисбаланс класов в таргете. (но не oversampling, это скользкая дорожка, через class_weight)


    

- полезно настраивать векторайзеры (тут пригодится pipeline)


  
    
- подобрать лучшие гиперпараметры с использованием кроссвалидации (тут пригодится GridSearchCV) Done!
    
    
    

- сгенерировать новые фичи, например  например посчитать число слов в тексте, длину слов итп итд. Или с помощью [тематического моделирования](https://pythobyte.com/python-for-nlp-topic-modeling-8fb3d689/) / использовать ембединги слов, учитывающие семантику, например [word2vec](https://radimrehurek.com/gensim/models/word2vec.html)) 
    
    
    

- попробовать другие модели. проект своеобразный выбор между вычислительными ограничениями (много примеров, расчеты могут затянуться) и задачей получить хорошую метрику (как это и бывает на практике), поэтому советовать "тяжелые", но мощные модели, чтобы у тебя все окончательно не повисло не буду (хотя есть вариант попробовать сделать на GPU).  А вот попробовать простые модели: SVC, NBC, логистическая регрессия, которые хорошо отработают с разряженными матрицами, могу. Простые модели - зато используем весь датасет. 



</div>


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:

В конце проекта принято кратко описывать все проделанные шаги и полученные результаты. Зачем это нужно - когда проект захочет посмотреть будущий работодатель, у него может не быть времени на подробный разбор кода. Вероятнее всего он бегло просмотрит код, но захочет изучить результат, который будет в общем выводе. Поэтому все же советую написать общий вывод пообьемней: добавить пару слов о данных, работе с ними, о моделях, метриках
<div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b>
Добавлен вывод 
</div>


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>



Сергей, у тебя старательно выполненная работа, все четко, осмысленно. 
    
    
    
    
Замечания на будущее:
    
- Комментарии к коду отсутствуют - коллеги могут не понять хода твоих мыслей. 

    
    
    
Я оставил небольшие советы и вопросики (если есть время и желание можешь воспользоваться/ответить).
    
    
    
    
Обязательное к исправлению:






- WordNetLemmatizer используем с POS - тег 





- .astype('U') лишнее, стоит экономить ресурсы, иначе может даже ядро обрушиться






- не забываем о random_state



- общий вывод - лицо проекта стоит сделать пообьемней   


    
Жду исправлений, для принятия проекта. Если какие то вопросы, то сразу спрашивай ) 


</div>



<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV2</b></font>

    
Глеб у меня лемматизация запустилась (в первый раз что то выскочила, потом все запустилось), сейчас трудно уже сказать что было не так. Если опять появится шли обратно, разберемся.

    
    
Что осталось из красного:

    
-  не забываем о random_state при инициализации модели

    
    
На связи    

</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b>
 добавил random_state
</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b>
 Марат, большое спасибо за советы и комментарии. Я вижу и обязательно к ним вернусь. просто сейчас плохо со временим из-за переезда  
</div>


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV3</b></font>



Глеб, тобой исправлено красное, желтое не тронуто - твое право ) Тем не менее надеюсь мои советы и вопросики будут полезны и ты узнаеешь что то новое.
Отличная работа. Желаю успехов в дальнейшей учебе!
    
</div>
